### DISCUSSION TITLE
How to access IHostingEnvironment, ContentRoot inside Program.cs Main method?

### DISCUSSION THREAD

**Discussion by tekbyts** [dotnet/aspnetcore#25695] [General]
This is my use case: 

Download `appsettings.json` from an external trust store and place it inside the web application project root directory. 
I should do this inside `Program.cs -> Main` method as the first thing to be done before `CreateWebHostBuilder` 
In order to achieve that, i need access to `IHostingEnvironment.ContentRootPath` 

Below is the code: i can declare a variable inside the `Main` method of type `IHostingEnvironment` but i am not sure what should i initialize it with 

```
public static void Main(string[] args)
{
    IHostingEnvironment env;
}
```

**Comment by davidfowl** [comment]
You can't access it inside of main. The host needs to be built before you can access anything there. The best you can do is set the content root.

**Reply by Tratcher** [reply]
What about doing it inside of a IHostBuilder.ConfigureAppConfiguration callback? As long as you registered that callback early enough...

**Reply by Tratcher** [reply]
```

            var host = new HostBuilder();
            host.ConfigureAppConfiguration((hostContext, config) =>
            {
                var hostEnvironment = hostContext.HostingEnvironment;
                // TODO: Copy the appsettings.json file...

                config.AddJsonFile("AppSettings.json", optional: false);
            });
```

The downside of this approach is that you can't use `Host.CreateDefaultBuilder` due to ordering issues. However, you can copy as much code for that from here as you need:
https://github.com/dotnet/runtime/blob/ac3595e47432b728ddccabdc280692ac0e9405f7/src/libraries/Microsoft.Extensions.Hosting/src/Host.cs#L56

**Reply by davidfowl** [reply]
That doesn't run until Build is called... I'm not 100% sure how this would work.

**Reply by Tratcher** [reply]
Sure, the download would happen during Build. As long as it happens before AddJsonFile is called it should work in this scenario.

**Reply by Tratcher** [reply]
Alternatively, you could do the download first into a temp location, and then copy it to the app on that TODO line.

Of course you could change the premise of the problem and write a config provider instead. No need to mess with the appsetting.json file. It would be something like the KeyVault provider.
https://github.com/dotnet/extensions/blob/release/3.1/src/Configuration/Config.AzureKeyVault/src/AzureKeyVaultConfigurationExtensions.cs

**Reply by davidfowl** [reply]
That's preferred I think.

**Comment by tekbyts** [comment]
Can you please provide a code sample?
